### ISSUE TITLE
[Bug]: assets_image_thumbnail_cache gets cleared and thumbnail is broken when a deferred thumbnail URL is requested twice

### ISSUE THREAD

**Issue by bitbirddev** [pimcore/pimcore#16486]
### Pimcore version

11.x

#### 1. Add Frontend Prefixes
```yaml
pimcore:
    assets:
        frontend_prefixes:
            source: "%env(CDN_DOMAIN)%/assets"
            thumbnail: "%env(CDN_DOMAIN)%/thumbnails"
```

#### 2. Clear/Truncate
- tmp_store
- assets_thumbnail_cache
- Temporary Files (via Button in Admin->Settings->Cache->Clear temporary files

#### 3. Create a template with an image-tag 
`{{ pimcore_image('test', {'thumbnail': {'width': 200, 'height': 200}}) | raw }}`

___we now have to make some request-roundtrips to get to the problem___

When you first open your template in the frontend, the image tag should look similar to this

```html
<picture>
	<source srcset="/3/image-thumb__3___auto_ea796264b15fae954fcc6cb999b3b659/testimage.e347a3c5.webp 1x, /3/image-thumb__3___auto_ea796264b15fae954fcc6cb999b3b659/testimage@2x.e347a3c5.webp 2x" width="200" height="200" type="image/webp">
	<source srcset="/3/image-thumb__3___auto_ea796264b15fae954fcc6cb999b3b659/testimage.e347a3c5.jpg 1x, /3/image-thumb__3___auto_ea796264b15fae954fcc6cb999b3b659/testimage@2x.e347a3c5.jpg 2x" width="200" height="200" type="image/jpeg">
	<img src="/3/image-thumb__3___auto_ea796264b15fae954fcc6cb999b3b659/testimage.e347a3c5.jpg" width="200" height="200" alt="" loading="lazy" srcset="/3/image-thumb__3___auto_ea796264b15fae954fcc6cb999b3b659/testimage.e347a3c5.jpg 1x, /3/image-thumb__3___auto_ea796264b15fae954fcc6cb999b3b659/testimage@2x.e347a3c5.jpg 2x">
</picture>
```

In my case my browser choose to load the .webp-Version of the srcset:

<img width="1005" alt="image" src="https://github.com/pimcore/pimcore/assets/15128942/59703a2a-019f-4736-b33e-144a6a802471">

So after the first request ___ONE___ thumbnail is generated 

![image](https://github.com/pimcore/pimcore/assets/15128942/c689072e-6b43-41f4-974b-0948cdb2e42d)

while the other 3 thumbnails are deferred and their configs saved to the tmp_store for later

![image](https://github.com/pimcore/pimcore/assets/15128942/41f58055-7a64-4e4a-aa62-495830eabaae)

everything is fine for now. now reload the page. as you can see, your picture-tag has changed. the thumbnail that was generated in the first request got his frontend_prefixes now and thus is loaded via S3 in my case.

```html
<picture>
	<source srcset="/3/image-thumb__3___auto_ea796264b15fae954fcc6cb999b3b659/testimage.e347a3c5.webp 1x, https://pimcore-boilerplate.s3.eu-central-1.amazonaws.com/thumbnails/3/image-thumb__3___auto_ea796264b15fae954fcc6cb999b3b659/testimage@2x.e347a3c5.webp 2x" width="200" height="200" type="image/webp">
	<source srcset="/3/image-thumb__3___auto_ea796264b15fae954fcc6cb999b3b659/testimage.e347a3c5.jpg 1x, /3/image-thumb__3___auto_ea796264b15fae954fcc6cb999b3b659/testimage@2x.e347a3c5.jpg 2x" width="200" height="200" type="image/jpeg">
	<img src="/3/image-thumb__3___auto_ea796264b15fae954fcc6cb999b3b659/testimage.e347a3c5.jpg" width="200" height="200" alt="" loading="lazy" srcset="/3/image-thumb__3___auto_ea796264b15fae954fcc6cb999b3b659/testimage.e347a3c5.jpg 1x, /3/image-thumb__3___auto_ea796264b15fae954fcc6cb999b3b659/testimage@2x.e347a3c5.jpg 2x">
</picture>
```

your database tables didn't change, `tmp_store` and `assets_image_thumbnails_cache` look just as in the screenshot above

### The final part

- choose a URL of your srcset that wasn't requested from your browser (the ones without fronted_prefix)
- Right click and open the image in a new Tab (you should see the image)
- Now Reload the tab => ___THE IMAGE IS GONE___
<img width="459" alt="SCR-20240118-bwhb" src="https://github.com/pimcore/pimcore/assets/15128942/ebd51dfa-5643-442b-8c7d-a23c355bf1c4">

Not only the image is gone, also your `assets_image_thumbnails_cache` is cleared. 

***Since the `tmp_store`-Entry for this specific thumbnail was deleted after the thumbnail was created and since the corresponding entry in `assets_image_thumbnails_cache` is now also missing, the thumb can never be generated ever again.***

![SCR-20240118-bxcs](https://github.com/pimcore/pimcore/assets/15128942/0567f3cc-2d22-4f1b-962b-cfd1313bc70f)

And the worst part is that for this image the `assets_image_thumbnails_cache` won't ever get filled again because the files still exist on the disk

### Why is this Bad
- loading a deferred thumbnail url twice wipes out all your `assets_image_thumbnails_cache`-Entries for this image
- it also breaks this specific image-thumbnail
- when using the full page cache, chances are very high that an deferred thumbnail url is cached => high chances that this url is requested twice
- chances are high that you don't see your missing images because your browser requests another version of the thumb. your users might see broken images even everything seems ok on your side

you might first think this is an edge-case and doesn't really happen but it happens all the time. i've noticed this in production for a while now. don't forget that not only users visit our sites but also a lot of crawlers. 

### What else
- it would be nice to have a console command to generate thumbs for entries in the `tmp_store`
- the full page cache should only be activated if no deferred thumbnail urls are found in the response (if `frontend_prefixes` are enabled)

### Actual Behavior

- requesting the deferred thumbnail url twice clears all entries for this image in `assets_image_thumbnail_cache`
- also the thumbnail can't be generated again now because its entry in `tmp_store` is missing now
- `frontend_prefixes` won't be added now for any of the thumbnails of the image

### Expected Behavior

- calling the deferred URL should always display an image thumb, even if its called 2x, 3x, 4x
- calling the deferred thumbnail URL should not wipe out other image thumbnails in `assets_image_thumbnails_cache` of this image
- thumbnail generation shouldn't be broken after calling a deferred thumbnail url twice

**Comment by micru** [comment]
We encounter the same problem!

**Comment by micru** [comment]
Thanks @bitbirddev for the great explanation and description on how to reproduce the issue. I can confirm the steps!

**Comment by khusseini** [comment]
We have the same problem. Thanks @bitbirddev for the detailed report.

**Comment by bitbirddev** [comment]
@micru @khusseini good to know that i'm not the only one!

**Comment by brusch** [comment]
@bitbirddev from what you've written, I'm pretty sure that #16467 fixes your issue. Could you please give it a try? 
@micru @khusseini

**Comment by brusch** [comment]
Btw. this should only happen when having a dynamic thumbnail configuration, like 
```twig
{{ pimcore_image('test', {'thumbnail': {'width': 200, 'height': 200}}) | raw }}
```

You shouldn't encounter this bug when using thumbnail configurations, like: 
```twig
{{ pimcore_image('test', {'thumbnail': 'exampleThumbConfig'}) | raw }}
```

The latter should be the preferred one, because it takes away a lot of overhead (`TmpStore` ...)

**Comment by bitbirddev** [comment]
> Btw. this should only happen when having a dynamic thumbnail configuration, like
> 
> ```
> {{ pimcore_image('test', {'thumbnail': {'width': 200, 'height': 200}}) | raw }}
> ```
> 
> You shouldn't encounter this bug when using thumbnail configurations, like:
> 
> ```
> {{ pimcore_image('test', {'thumbnail': 'exampleThumbConfig'}) | raw }}
> ```
> 
> The latter should be the preferred one, because it takes away a lot of overhead (`TmpStore` ...)

it happens as soon as you select a specific area of an image in the {{ pimcore_image }} editable.

**Comment by brusch** [comment]
@bitbirddev yeah, you're right, this is also a case of dynamic thumbnail generation 😉 
Any feedback about https://github.com/pimcore/pimcore/pull/16467  fixing the issue?

**Comment by bitbirddev** [comment]
> @bitbirddev yeah, you're right, this is also a case of dynamic thumbnail generation 😉 Any feedback about #16467 fixing the issue?

@brusch #16467 does only fix one part of the issue. `assets_image_thumbnail_cache` still gets cleared for the thumbnail if a deferred thumbnail url is requested twice - and the entries never get added again.

on the second request all entries in `assets_image_thumbnail_cache` for the image are deleted here

https://github.com/pimcore/pimcore/blob/48d2cad66f027888d7d1bb1acd94cc9abed4ab77/models/Asset/Service.php#L503

the only place where entries do get added to the thumbnail cache is here 

https://github.com/pimcore/pimcore/blob/48d2cad66f027888d7d1bb1acd94cc9abed4ab77/models/Asset/Image/Thumbnail/Processor.php#L343-L346

but since the thumbnail still exists in the storage on the second request, this code is never reached again.

see #16523 as quick workaround. 

its not perfekt as requesting the deferred url twice still gives you that [filetype-not-supported.svg](https://github.com/pimcore/pimcore/blob/48d2cad66f027888d7d1bb1acd94cc9abed4ab77/bundles/CoreBundle/src/Controller/PublicServicesController.php#L59) but the image generation + frontend_prefixes are not completely broken after that.

**Comment by brusch** [comment]
@bitbirddev can you please try #16529 if it fixes your issue? Thanks again! 😊

**Comment by bitbirddev** [comment]
> @bitbirddev can you please try #16529 if it fixes your issue? Thanks again! 😊

still testing but looks really good so far! will keep you updated!


- do you think it would be good to have a console command to generate all the thumbnails for all thumbnail entries in tmp_store?
- should me introduce an option to only enable the full page output cache if all thumbnails are generated for the page? like a regex that check for deferred thumbnail urls in the response before it gets cached...

**Comment by bitbirddev** [comment]
@brusch its a lot better know but the bug i fixed in #16523 still exists.

1. create a twig template with
```twig
{% set image = pimcore_image('test', {'thumbnail': {'width': 200, 'height': 200}}) %}
{{ image }}
```

2. clear all thumbnails of the image of your editable
<img width="813" alt="image" src="https://github.com/pimcore/pimcore/assets/15128942/a3eff6c2-62a9-4f3f-9156-072321091773">


now open your template in your browser and manually copy the paths of the first 2 source urls and open them in a new window.

reload the page. you should see something like this
<img width="852" alt="image" src="https://github.com/pimcore/pimcore/assets/15128942/85690743-8b32-40f5-b162-b00dc5383562">

two of our urls now have the frontend prefixes added and `assets_image_thumbnail_cache` looks like this
![image](https://github.com/pimcore/pimcore/assets/15128942/7a8ee2c6-b977-4482-8ed7-0a34f4f49ca8)

now copy an url ***without*** frontend prefixes and request it ***twice in parallel***. In most of the cases i'm not fast enough with my mouse so i'm using this shell command 
```bash
 parallel curl https://pimcore-boilerplate.test/3/image-thumb__3___auto_ea796264b15fae954fcc6cb999b3b659/testimage.e347a3c5.jpg ::: `seq 1 2`
```

after that request reload the page again. it should look like this

<img width="866" alt="image" src="https://github.com/pimcore/pimcore/assets/15128942/5c644449-f7fd-4037-b280-c0c492025348">

we've lost our frontend_prefixes for the first 2 thumbnails. the entries are now also missing in the thumbnail_cache

![image](https://github.com/pimcore/pimcore/assets/15128942/9cd067b7-5739-4d2d-89ae-1849db8c4828)

you can now request the first 2 thumbnail urls again and as often as you want, you'll never get your frontend_prefixes back ever again.

**Comment by brusch** [comment]
@bitbirddev Thanks for the feedback and further clarifying the issue. 
I've just updated my PR #16529 with a [commit ](https://github.com/pimcore/pimcore/pull/16529/commits/95cd450be43672ad3a5ac4f4e2842aabf52c1cf2) that should fix this issue as well 😊 

Can you please try it again on your side?

**Comment by bitbirddev** [comment]
@brusch i think we're finally there! tried everything but couldn't break it this time. thank you so much! i can finally trust my thumbnails again :P hope this one gets merged soon so i can update my projects. thanks again!

**Comment by brusch** [comment]
@bitbirddev Thanks for your feedback 😊

**Comment by bitbirddev** [comment]
@brusch anytime. thanks for taking the time to look into it!

**Comment by brusch** [comment]
Fixed by https://github.com/pimcore/pimcore/pull/16529
